satpy/readers/ami_l1b.py

```diff
@@ -0,0 +1,230 @@
+from satpy.readers.file_handlers import BaseFileHandler
+from satpy import CHUNK_SIZE
+
+logger = logging.getLogger(__name__)
+
+PLATFORM_NAMES = {
+    'GK-2A': 'GEO-KOMPSAT-2A',
+    'GK-2B': 'GEO-KOMPSAT-2B',
+}
+
+
+class AMIL1bNetCDF(BaseFileHandler):
+    """Base reader for AMI L1B NetCDF4 files."""
+
+    def __init__(self, filename, filename_info, filetype_info,
+                 calib_mode='PYSPECTRAL', allow_conditional_pixels=False):
+        """Open the NetCDF file with xarray and prepare the Dataset for reading."""
+        super(AMIL1bNetCDF, self).__init__(filename, filename_info, filetype_info)
+        self.nc = xr.open_dataset(self.filename,
+                                  decode_cf=True,
+                                  mask_and_scale=False,
+                                  chunks={'dim_image_x': CHUNK_SIZE, 'dim_image_y': CHUNK_SIZE})
+        self.nc = self.nc.rename({'dim_image_x': 'x', 'dim_image_y': 'y'})
+
+        platform_shortname = self.nc.attrs['satellite_name']
+        self.platform_name = PLATFORM_NAMES.get(platform_shortname)
+        self.sensor = 'ami'
+        self.allow_conditional_pixels = allow_conditional_pixels
+        calib_mode_choices = ('FILE', 'PYSPECTRAL')
+        if calib_mode.upper() not in calib_mode_choices:
+            raise ValueError('Invalid calibration mode: {}. Choose one of {}'.format(
+                calib_mode, calib_mode_choices))
+        self.calib_mode = calib_mode.upper()
+
+    @property
+    def start_time(self):
+        """Get observation start time."""
+        base = datetime(2000, 1, 1, 12, 0, 0)
+        return base + timedelta(seconds=self.nc.attrs['observation_start_time'])
+
+    @property
+    def end_time(self):
+        """Get observation end time."""
+        base = datetime(2000, 1, 1, 12, 0, 0)
+        return base + timedelta(seconds=self.nc.attrs['observation_end_time'])
+
+    def get_area_def(self, dsid):
+        """Get area definition for this file."""
+        a = self.nc.attrs['earth_equatorial_radius']
+        b = self.nc.attrs['earth_polar_radius']
+        h = self.nc.attrs['nominal_satellite_height'] - a
+        lon_0 = self.nc.attrs['sub_longitude'] * 180 / np.pi  # it's in radians?
+        cols = self.nc.attrs['number_of_columns']
+        rows = self.nc.attrs['number_of_lines']
+        obs_mode = self.nc.attrs['observation_mode']
+        resolution = self.nc.attrs['channel_spatial_resolution']
+
+        cfac = self.nc.attrs['cfac']
+        coff = self.nc.attrs['coff']
+        lfac = self.nc.attrs['lfac']
+        loff = self.nc.attrs['loff']
+        bit_shift = 2**16
+        area_extent = (
+            h * np.deg2rad((0 - coff - 0.5) * bit_shift / cfac),
+            -h * np.deg2rad((0 - loff - 0.5) * bit_shift / lfac),
+            h * np.deg2rad((cols - coff + 0.5) * bit_shift / cfac),
+            -h * np.deg2rad((rows - loff + 0.5) * bit_shift / lfac),
+        )
```

Comment:
This looks really similar to the hrit code, any possibility to factorize these ?
https://github.com/pytroll/satpy/blob/master/satpy/readers/hrit_base.py#L262-L277

Reply:
It does look pretty similar doesn't it. That's probably because these L1b files are based off of "uhrit" files which are probably similar in data structure to hrit files. I'm going to say that I don't want to make this refactor in this PR. We can make it an issue if you'd like, but there is a lot in that module that has nothing to do with these AMI files. We could move the method to another module, but it is also pretty specific to hrit, right?

Reply:
Yes, it's hrit specific, but I don't see any reason not to factorize. An issue would be the least :)

Reply:
Just a note, it's not HRIT specific but Geostationary specific. The AGRI reader uses similar code. Perhaps this type of thing could be added to a `geo_utils.py` or similar file?

Reply:
@simonrp84 But the `2**16` part isn't geostationary specific, right? We don't need these types of calculations for ABI because we have the X/Y radian coordinates already.

Reply:
Yes, @djhoese that's right. But ABI is a bit of an exception, we could do it the same way as the other geosats - but why bother with the extra computation :) 
To my knowledge all the other geosats need this calculation.

Reply:
ABI is CF compliant. Hopefully future satellite data files (at least for higher level processing output) can also be CF compliant.

Reply:
Agreed!